cognify/optimizer/core/sh.py

```diff
@@ -0,0 +1,116 @@
+from typing import Callable
+import logging
+from concurrent.futures import (
+    ThreadPoolExecutor,
+    as_completed,
+)
+from concurrent.futures import Future
+
+from cognify.optimizer.evaluator import (
+    EvaluationResult,
+    GeneralEvaluatorInterface,
+)
+from cognify.optimizer.core.common_stats import CommonStats
+from cognify.optimizer.core.flow import TopDownInformation
+
+logger = logging.getLogger(__name__)
+
+class SuccessiveHalving:
+    """Policy for successive halving resource allocation
+    
+    If a layer (A) adopts this policy, it will allocate resources for the next layer (B)
+    
+    layer A will first propose `n` trials, then
+    
+    **SH routine**:
+    1. each remaining trial at layer B will have a fixed step_budget = `r_i`
+    2. after all trials are evaluated, unpromising trials will be pruned (top `n = n/prune_rate` will be kept)
+    3. repeat from step 1 until no trials left or iteration limit reached
+    """
+    def __init__(
+        self,
+        prune_rate: int,
+        num_SH_iter: int,
+        initial_step_budget: int,
+        hierarchy_level: int,
+        next_layer_factory: Callable[[], GeneralEvaluatorInterface],
+        selected_runs: list[TopDownInformation],
+    ):
+        self.prune_rate = prune_rate
+        self.num_SH_iter = num_SH_iter
+        self.initial_step_budget = initial_step_budget
+        self.hierarchy_level = hierarchy_level
+        self.selected_runs: list[TopDownInformation] = selected_runs
+        self.can_continue = [i for i in range(len(selected_runs))]
+        self.num_inner_trials = [0] * len(selected_runs)
+        self._next_layer_factory = next_layer_factory
+    
+    def _evaluate_one_config(self, i) -> EvaluationResult:
+        _next_layer = self._next_layer_factory()
+        # NOTE: frac not needed bc next layer will never be evaluator
+        result = _next_layer.evaluate(self.selected_runs[i])
+        return result
+
+    def _halving_jointly(self, indicators, not_converged, n_left):
+        """Round-Robin in each dimension
+        """
+        # NOTE: indicators all smaller the better
+        groups_by_metric = list(map(list, zip(*indicators)))
+        sorted_indices = [sorted(not_converged, key=lambda x: group[x]) for group in groups_by_metric]
+        # round robin select from top of each list
+        # avoid duplicates
+        left_to_run = set()
+        from_group = 0
+        while len(left_to_run) < n_left:
+            # if all indices are empty, break
+            if not any(len(sorted_indices[i]) > 0 for i in range(len(groups_by_metric))):
+                break
+            if len(sorted_indices[from_group]) > 0:
+                left_to_run.add(sorted_indices[from_group].pop(0))
+            from_group = (from_group + 1) % len(groups_by_metric)
```

Comment:
when selecting from RR, the entries on different list could be the same one, right? do you check that and don't count towards the n_left?

Reply:
Right, I use set for `left_to_run` to avoid duplication.